### ISSUE TITLE
incoherent-version-in-changelog is factoring .%dist in %{release}

### ISSUE THREAD

**Issue by brianjmurrell** [rpm-software-management/rpmlint#856]
I am getting a `W: incoherent-version-in-changelog 1.15.0~rc3-2 ['1.15.0~rc3-2.suse.lp153', '1.15.0~rc3-2.suse.lp153']` warning.

This is due to `Release: 2%{?dist}` in the specfile.  But I cannot include either `%{dist}` or even `%{release}` in the changelog version or else I get `W: macro-in-%changelog %{release}`.  I certainly cannot hard code a given distro's `%{dist}` value either.

It seems like https://github.com/rpm-software-management/rpmlint/blob/b749f3ada9fb708ea3dc3e3de515e4733e892496/rpmlint/checks/TagsCheck.py#L549-L562 needs to account for a possible `.%{?dist}` in `Release:`.

**Comment by marxin** [comment]
> This is due to `Release: 2%{?dist}` in the specfile.

It seems strange adding `%{dist}` to `Release`.

> But I cannot include either `%{dist}` or even `%{release}` in the changelog version or else I get `W: macro-in-%changelog %{release}`. I certainly cannot hard code a given distro's `%{dist}` value either.

Well, I can imagine supporting any of the 2 macros in `%changelog` but I would like to see your use-case?

**Comment by brianjmurrell** [comment]
> It seems strange adding `%{dist}` to `Release`.

To just pick a specfile at random:

https://src.fedoraproject.org/rpms/autofs/blob/rawhide/f/autofs.spec#_15

I think you will find that all Fedora, RHEL (and therefore all EL downstreams) and SUSE specfiles do that.

> Well, I can imagine supporting any of the 2 macros in `%changelog` but I would like to see your use-case?

As above?  But honestly, I'd rather not use `%{release}` as it defeats the purpose of the changelog and having to add `%{?dist}` to every changelog entry just seems silly.

It seems far more useful for `expected` to just allow for `%{version}-%{release}` with `.%{dist}` removed.  But yes, how do you know what `%{dist}` is for a SUSE RPM when running `rpmlint` on say, a Fedora machine.  I can only suppose you need to keep a list of common `%{dist}`s.

**Comment by marxin** [comment]
> To just pick a specfile at random:
> 
> https://src.fedoraproject.org/rpms/autofs/blob/rawhide/f/autofs.spec#_15
> 
> I think you will find that all Fedora, RHEL (and therefore all EL downstreams) and SUSE specfiles do that.
> 

All right, so then let's allow `%{dist}` in `Release` then. Do you want to come up with a pull request?

**Comment by marxin** [comment]
Please upload somewhere RPM files so that I can observe the difference.

**Comment by brianjmurrell** [comment]
Hrm.  I don't really have anywhere I can upload them to.

If you use mock to build https://github.com/daos-stack/dpdk/blob/1231cac43173877aae6b152ded5523d80809bac9/dpdk.spec on each of opensuse-leap-15.3-x86_64 and rocky+epel-8-x86_64 you will get RPMs you can compare.

**Comment by marxin** [comment]
> Hrm. I don't really have anywhere I can upload them to.

Please zip them and attach them to this Issue.

**Comment by marxin** [comment]
Can you please attach the problematic RPM file?

**Comment by brianjmurrell** [comment]
[libfabric-1.15.1-3.suse.lp153.x86_64.rpm.tar.gz](https://github.com/rpm-software-management/rpmlint/files/9489801/libfabric-1.15.1-3.suse.lp153.x86_64.rpm.tar.gz)

**Comment by marxin** [comment]
Thanks for the file.

So can you please experiment with something like:

```patch
diff --git a/rpmlint/checks/SpecCheck.py b/rpmlint/checks/SpecCheck.py
index 997d974c..739840be 100644
--- a/rpmlint/checks/SpecCheck.py
+++ b/rpmlint/checks/SpecCheck.py
@@ -436,7 +436,7 @@ class SpecCheck(AbstractCheck):
                                          '%%changelog: %s' % deptoken)
                 for match in self.macro_regex.findall(line):
                     res = re.match('%+', match)
-                    if len(res.group(0)) % 2 and match != '%autochangelog':
+                    if len(res.group(0)) % 2 and match not in ('%autochangelog', '%release', '%dist'):
                         self.output.add_info('W', pkg, 'macro-in-%changelog', match)
             else:
                 if not depscript_override:
```

which should not warn about `%release` nor `%dist` macros in `%changelog`.

**Comment by Conan-Kudo** [comment]
The `%release` macro should never be in the changelog, nor should `%dist`.

**Comment by Conan-Kudo** [comment]
@brianjmurrell In the case of writing out EVRs in the changelog entry, you should omit the `%dist` macro, because that value changes depending on who it is built for.

**Comment by brianjmurrell** [comment]
@Conan-Kudo Indeed, I completely agree.  While it's been a while since I filed this ticket, because I generally agree with you I can only imagine that I tried including `%{dist}` to try to appease some other `rpmlint` error.  I will try to revisit this.

**Comment by brianjmurrell** [comment]
> ```diff
> if len(res.group(0)) % 2 and match not in ('%autochangelog', '%release', '%dist'):
> ```

But per @Conan-Kudo (and I agree with him), I shouldn't actually be putting the `%dist` macro into the `%changelog` entry, so this solution seems to simply allow the hacky work-around rather than addressing the root problem of having the `%dist` in the `Release:` being factored into the comparison of the version in the `%changelog`.

That said, the work-around does work.  It just doesn't seem to be the correct solution where the correct solution, I would think, is to disable the factoring of `%dist` into `Release:` when comparing the version with the one in the `%changelog`.

**Comment by brianjmurrell** [comment]
Any further thoughts given the last couple of comments?

**Comment by brianjmurrell** [comment]
What is the workaround exactly?

**Comment by marxin** [comment]
The one mentioned in https://github.com/rpm-software-management/rpmlint/issues/856#issuecomment-1252382833 ?

**Comment by brianjmurrell** [comment]
Do you mean using `%{dist}` in the `%changelog` entry?  It seems count-productive to break the data just to appease a broken validation tool.

I think I'd sooner add an ignore in the `rpmlintrc` than to start breaking the real-world data just to appease the broken validator.

So I guess that is what I will do until this issue is fixed.

Thanks for all of the input.

**Comment by marxin** [comment]
> I think I'd sooner add an ignore in the `rpmlintrc` than to start breaking the real-world data just to appease the broken validator.

Yes, that would be the easiest approach.

Thanks for the discussion.
